modules/GenericDataParser/src/server/generic_data_parser.R

```diff
@@ -2908,9 +2908,330 @@ splitOnSemicolon <- function(x) {
+        SSE <- NA
+        SSR <- NA
+        SST <- NA
+        rSquared <- NA
+        if(!is.na(renderingOptions$fct) && !missingParameters) {
+          tmp <- as.data.frame(fixedParams[[1]])
+          fct <- eval(parse(text=paste0('function(x) ', renderingOptions$fct)))
+          plotLog <- "x"
+          for(i in 1:ncol(tmp)) {
+              assign(names(tmp)[i], tmp[,i])
+          }
+          goodPoints <- racas::filterFlaggedPoints(points[[1]], returnGood = TRUE)
+          SSE <- sum((fct(goodPoints$dose) - goodPoints$response)^2)
+          SSR <- sum((fct(goodPoints$dose) - mean(goodPoints$response))^2)
+          SST <- SSR + SSE
+          rSquared <- SSR/SST
+        }
+
+        # Apply parsed params to fit data
+        # This applies color preferences, plot limits etc. to the fit data and parsed params objects
+        # It also returns the curve fit data as points and params as the renderer requires
+        output <- racas::applyParsedParametersToFitData(fitData, parsedParams, protocolDisplayValues)
+        data <- output$data
+        parsedParams <- output$parsedParams
+
+        # Plot the curve and get the base64 encoded text of the image
+        t <- tempfile()
+        suppressWarnings(racas::plotCurve(curveData = data$points, params = data$parameters, drawCurve = TRUE, logDose = parsedParams$logDose, logResponse = parsedParams$logResponse, outFile = t, ymin=parsedParams$yMin, ymax=parsedParams$yMax, xmin=parsedParams$xMin, xmax=parsedParams$xMax, height=parsedParams$height, width=parsedParams$width, showGrid = parsedParams$showGrid, showAxes = parsedParams$showAxes, labelAxes = parsedParams$labelAxes, showLegend=parsedParams$legend, mostRecentCurveColor = parsedParams$mostRecentCurveColor, axes = parsedParams$axes, plotColors = parsedParams$plotColors, curveLwd=parsedParams$curveLwd, plotPoints=parsedParams$plotPoints, xlabel = parsedParams$xLab, ylabel = parsedParams$yLab))
+        imgTxt <- RCurl::base64Encode(readBin(t, "raw", file.info(t)[1, "size"]), "txt")
+        unlink(t)
+        
+        # Apply the goodness of fit rules and return the table row with stats
+        trStyleClass = ""
+        statsCells <- ""
+        for(s in 1:length(goodnessOfFitParameters)) {
+            tdStyleClass <- ""
+            title <- ""
+            statThresholds <- goodnessOfFitParameters[[s]]
+            stat <- names(goodnessOfFitParameters)[s]
+            # Get the stat variable value from the environment
+            statValue <- get(stat)
+            statValueText <- format(statValue, digits=3)
+            warnings <- list()
+            errors <- list()
+            for(r in 1:length(statThresholds)) {
+                threshold <- statThresholds[[r]]
+                class <- names(statThresholds)[r]
+                if(is.na(statValue) || !is.na(threshold$value) && !is.na(threshold$operator) && eval(parse(text = paste0('statValue ', threshold$operator, ' threshold$value')))) {
```

Comment:
This whole section for each statThreshold is pretty dense. We're already 2 or 3 levels nested in a for loop, and we're doing both classification of curves and HTML formatting. 
I'd suggest shifting this portion out into a helper function that takes in the necessary info and yields an error + errorClass + title. That will keep this big nested for loop simpler and easier to read.

If you wanted to go a step further you could have this whole loop just return lists or tables of data (without HTML formatting) then do all the `<td>`, `<tr>` HTML formatting below.

Reply:
I did some of this.  I still have a lot of paste for the rows but I did implement some Markdown stuff which makes it a little cleaner looking.